packages/microcosm/src/domain-engine.js

```diff
@@ -76,28 +69,50 @@ class DomainEngine {
     return this.registry[type]
   }
 
-  add(key: string | KeyPath, config: *, options?: Object) {
+  add(key: string, config: *, options?: Object) {
     console.assert(
       !options || options.constructor === Object,
-      'addDomain expected a plain object as the second argument.',
-      'Instead got',
-      get(options, 'constructor.name', 'Unknown')
+      'addDomain expected a plain object as the third argument.'
     )
 
+    console.assert(key && key.length > 0, 'Can not add domain to root level.')
+
     let deepOptions = merge(
       this.repo.options,
       config.defaults,
       { key },
       options
     )
 
-    let keyPath: KeyPath = castPath(key)
     let domain: Domain = createOrClone(config, deepOptions, this.repo)
 
-    this.domains.push([keyPath, domain])
-
-    // Reset the registry
-    this.registry = {}
+    this.domains[key] = domain
+
+    this.lifecycle[RESET.toString()].push({
+      key: key,
+      scope: domain,
+      local: true,
+      steps: [
+        (state: *, data: any) => {
+          return data[key] !== undefined
+            ? data[key]
+            : result(domain, 'getInitialState')
+        }
+      ]
+    })
+
+    this.lifecycle[PATCH.toString()].push({
+      key: key,
+      scope: domain,
+      local: true,
+      steps: [
+        (state: *, data: any) => {
+          return data[key] !== undefined ? data[key] : state
+        }
+      ]
+    })
```

Comment:
can you explain what's up with `RESET.toString()` and `PATCH.toString()`? And what these two pushes do?

Reply:
Yep. 

`PATCH.toString()` is to tell Flow that the underlying value is a string. This is _lame_.

These two pushes replace the MetaDomain as the method of modifying data after `repo.reset` or `repo.patch` fire.

For reset:

Update the value for the key the domain is managing if the key is defined in the data payload. Otherwise fall back to the initial state. So `repo.reset()` reverts to initial state, but `repo({ key: 'value' })` sets a specific value.

For patch:

The same is true, however the fallback is the _current_ value, rather than the initial state.